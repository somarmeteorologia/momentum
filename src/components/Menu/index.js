import React, { useContext, useState, memo } from 'react'
import styled, { ThemeContext } from 'styled-components'
import { theme } from 'styled-tools'
import PropTypes from 'prop-types'
import { motion } from 'framer-motion'

import { Icon } from '@components/Icon'

const Container = styled(motion.div)`
  height: 100vh;
  background-color: ${theme('menu.bg.primary')};
  position: relative;
  display: block;
  box-shadow: ${theme('navigation.shadow.primary')};
`

const Toggleable = styled.div`
  width: 23px;
  height: 23px;
  cursor: pointer;
  border: 1px solid ${theme('menu.toggle.secondary')};
  background-color: ${theme('menu.toggle.primary')};
  display: flex;
  justify-content: center;
  align-items: center;
  border-radius: ${theme('border.radius.fifty')};
  position: absolute;
  right: -10px;
  top: 122px;
  z-index: ${theme('zindex.above')};

  &:hover {
    background-color: ${theme('menu.toggle.tertiary')};
  }
`

const Handler = styled(motion.div)`
  width: 300px;
  height: 100%;
  box-sizing: content-box;
  padding-right: 20px;
  position: fixed;
  display: block;
  z-index: ${theme('zindex.above')};
  top: 0;
  left: 0;
  touch-action: pan-y;
`

const Overlay = styled(motion.div)`
  width: calc(100% - 300px);
  height: 100%;
  position: absolute;
  z-index: ${theme('zindex.above')};
  right: 0;
  top: 0;
  background-color: transparent;
`

const Content = styled(motion.div)``

const Menuable = memo(
  ({ children, onChange, isOpen, setOpen, draggable, ...props }) => {
    const [hasMenuHover, setHasMenuHover] = useState(false)
    const [hasToggleHover, setHasToggleHover] = useState(false)
    const { menu } = useContext(ThemeContext)

    return (
      <Container
        {...props}
        data-testid="menu"
        onMouseEnter={() => setHasMenuHover(true)}
        onMouseLeave={() => setHasMenuHover(false)}
      >
        {children}

        {!draggable && hasMenuHover && (
          <Toggleable
            data-testid="toggleable"
            onMouseEnter={() => setHasToggleHover(true)}
            onMouseLeave={() => setHasToggleHover(false)}
            onClick={() => {
              setOpen(!isOpen)
              onChange(!isOpen)
            }}
          >
            <Icon
              width={10}
              height={10}
              name={isOpen ? 'right' : 'left'}
              color={
                hasToggleHover ? menu.toggle.secondary : menu.toggle.tertiary
              }
            />
          </Toggleable>
        )}
      </Container>
    )
  }
)

export const Menu = memo(
  ({ isOpen, setOpen, onChange, children, draggable }) => {
    const { animations } = useContext(ThemeContext)

    const onPan = (_, info) => setOpen(info.offset.x < 0)

    return (
      <>
        {draggable ? (
          <>
            <Handler
              initial={animations.visible}
              variants={{
                visible: {
                  x: 0
                },
                collapsed: {
                  x: -300
                }
              }}
              animate={isOpen ? animations.collapsed : animations.visible}
              onPan={onPan}
            >
              <Menuable
                isOpen={isOpen}
                draggable={draggable}
                setOpen={setOpen}
                onChange={onChange}
              >
                <Content
                  initial={animations.visible}
                  animate={isOpen ? animations.collapsed : animations.visible}
                  variants={{
                    visible: {
                      opacity: 1,
                      x: 0
                    },
                    collapsed: {
                      opacity: 0,
                      x: -300
                    }
                  }}
                >
                  {children}
                </Content>
              </Menuable>
            </Handler>

            {!isOpen && <Overlay onClick={() => setOpen(true)} />}
          </>
        ) : (
          <Menuable
            draggable={draggable}
            initial={animations.visible}
            animate={isOpen ? animations.collapsed : animations.visible}
            variants={{
              visible: {
                minWidth: '300px'
              },
              collapsed: {
                minWidth: '20px'
              }
            }}
            isOpen={isOpen}
            setOpen={setOpen}
            onChange={onChange}
          >
            <Content
              initial={animations.visible}
              animate={isOpen ? animations.collapsed : animations.visible}
              variants={{
                visible: {
                  opacity: 1,
                  x: 0
                },
                collapsed: {
                  opacity: 0,
                  x: -300
                }
              }}
            >
              {children}
            </Content>
          </Menuable>
        )}
      </>
    )
  }
)

Menu.propTypes = {
  onChange: PropTypes.func,
  isOpen: PropTypes.bool.isRequired,
  setOpen: PropTypes.func.isRequired,
  draggable: PropTypes.bool
}

Menu.defaultProps = {
  draggable: false,
  onChange: () => {}
}
