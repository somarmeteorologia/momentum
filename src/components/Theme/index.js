import { tokens } from '@components/Tokens'

const { font, colors, border, zindex, animations, shadow } = tokens
const { white, ebony, cornflower, red, green, yellow, gray, ciano } = colors

const withCommons = theme => ({
  colors,
  font,
  border,
  zindex,
  animations,
  shadow,
  ...theme
})

export const Theme = {
  dark: withCommons({
    text: {
      primary: white.hundred,
      secondary: ebony.eighty,
      tertiary: cornflower.fifty,
      quaternary: ebony.seventy,
      disabled: ebony.sixty
    },
    bg: {
      primary: ebony.zero,
      secondary: ebony.thirteen,
      tertiary: ebony.sixteen
    },
    icons: {
      primary: white.hundred,
      secondary: cornflower.fifty
    },
    button: {
      bg: {
        primary: ebony.twenty,
        stroke: 'transparent',
        flat: 'transparent',
        rounded: white.hundred,
        disabled: ebony.twenty,
        group: 'transparent'
      },
      text: {
        primary: white.hundred,
        stroke: ebony.sixty,
        flat: white.hundred,
        group: ebony.sixty,
        rounded: cornflower.fifty,
        disabled: ebony.sixty
      },
      border: {
        primary: 'none',
        stroke: `1px solid ${ebony.sixty}`,
        flat: 'none',
        rounded: 'none',
        disabled: 'none',
        group: `1px solid ${ebony.twenty}`
      },
      hover: {
        primary: { bg: cornflower.fifty, text: white.hundred },
        stroke: { border: `1px solid ${white.hundred}`, text: white.hundred },
        flat: { bg: cornflower.fifty, text: white.hundred },
        group: { bg: ebony.twenty, text: white.hundred },
        rounded: { bg: white.hundred, text: cornflower.fifty }
      },
      active: {
        primary: cornflower.thirty,
        stroke: ebony.thirteen,
        flat: cornflower.thirty,
        group: { bg: ebony.twenty, border: 'none' }
      }
    },
    slider: {
      bg: {
        on: cornflower.fifty,
        off: ebony.twenty,
        thumb: white.hundred
      },
      text: {
        primary: white.hundred,
        secondary: white.hundred
      },
      border: {
        thumb: `2px solid ${ebony.seventy}`
      }
    },
    drawer: {
      bg: {
        primary: ebony.twenty
      },
      text: {
        primary: white.hundred,
        secondary: ebony.sixty
      },
      border: {
        primary: `1px solid ${ebony.sixteen}`
      },
      icon: {
        primary: white.hundred,
        secondary: cornflower.fifty
      },
      shadow: {
        primary: 'none'
      }
    },
    accordion: {
      bg: {
        primary: ebony.twenty
      },
      text: {
        primary: white.hundred
      },
      icon: {
        primary: ebony.sixty,
        secondary: white.hundred
      }
    },
    tabs: {
      text: {
        on: white.hundred,
        off: ebony.seventy
      },
      bg: {
        primary: 'transparent',
        secondary: cornflower.fifty
      }
    },
    selector: {
      text: {
        primary: white.hundred,
        disabled: ebony.sixty
      },
      border: {
        on: `1px solid ${cornflower.fifty}`,
        off: `1px solid ${ebony.sixty}`,
        disabled: `1px solid ${ebony.forty}`
      },
      bg: {
        on: cornflower.fifty,
        off: ebony.twenty
      },
      icon: {
        primary: white.hundred,
        disabled: ebony.forty
      }
    },
    tooltip: {
      text: {
        primary: ebony.zero,
        secondary: white.hundred
      },
      bg: {
        primary: white.hundred,
        secondary: cornflower.fifty
      },
      header: {
        primary: ebony.ninetySix,
        secondary: ebony.twenty
      },
      shadow: {
        primary: 'none'
      }
    },
    snackbar: {
      bg: {
        primary: ebony.zero
      },
      text: {
        primary: white.hundred,
        secondary: ciano.fifty
      }
    },
    notification: {
      bg: {
        primary: white.hundred
      },
      text: {
        primary: ebony.zero
      },
      icon: {
        primary: white.hundred,
        secondary: ebony.zero
      },
      close: {
        primary: ebony.zero
      },
      state: {
        primary: cornflower.fifty,
        danger: red.fifty,
        warning: yellow.fifty,
        success: green.fifty
      }
    },
    field: {
      bg: {
        primary: ebony.thirteen,
        secondary: white.hundred,
        tertiary: ebony.twenty,
        active: ebony.forty,
        hover: cornflower.fifty
      },
      text: {
        primary: white.hundred,
        secondary: ebony.sixty,
        tertiary: ebony.twenty,
        quaternary: white.hundred,
        active: cornflower.fifty,
        selected: white.hundred,
        danger: red.fifty,
        disabled: ebony.forty
      },
      border: {
        primary: `1px solid ${ebony.twenty}`,
        secondary: `1px solid ${ebony.forty}`,
        tertiary: '1px solid transparent',
        danger: `1px solid ${red.fifty}`
      },
      shadow: {
        primary: '0 2px 4px rgba(0, 0, 0, 0.1)'
      },
      icon: {
        primary: ebony.sixty,
        secondary: ebony.forty,
        tertiary: ebony.sixty,
        active: white.hundred,
        success: green.fifty
      },
      arrow: {
        primary: ebony.sixty,
        selected: white.hundred
      }
    },
    datepicker: {
      text: {
        primary: ebony.zero,
        secondary: white.hundred,
        tertiary: ebony.sixty,
        off: ebony.seventy
      },
      bg: {
        primary: white.hundred,
        secondary: cornflower.fifty,
        tertiary: ebony.thirteen
      },
      icon: {
        primary: cornflower.fifty,
        secondary: white.hundred,
        tertiary: ebony.sixty
      },
      shadow: {
        primary: 'none'
      },
      border: {
        primary: `1px solid ${ebony.twenty}`
      }
    },
    navigation: {
      bg: {
        primary: ebony.thirteen,
        secondary: ebony.sixteen,
        tertiary: ebony.sixty,
        hover: white.hundred
      },
      separator: {
        primary: ebony.sixteen
      },
      text: {
        primary: white.hundred,
        hover: cornflower.fifty
      },
      shadow: {
        primary: 'none'
      }
    },
    menu: {
      bg: {
        primary: ebony.thirteen
      },
      toggle: {
        primary: ebony.sixteen,
        secondary: ebony.thirteen,
        tertiary: white.hundred
      }
    },
    player: {
      text: {
        primary: white.hundred,
        secondary: ebony.zero
      },
      bg: {
        primary: ebony.twenty,
        secondary: white.hundred,
        opacity: '0.8'
      },
      icon: {
        primary: white.hundred,
        secondary: ebony.twenty
      },
      progress: {
        on: white.hundred,
        off: gray.fifty
      },
      shadow: {
        primary: 'none'
      }
    },
    datatable: {
      text: {
        primary: white.hundred,
        hover: ebony.zero
      },
      bg: {
        primary: ebony.twenty,
        rowHover: white.hundred,
        columnHover: ebony.sixteen
      },
      border: {
        primary: `1px solid ${ebony.thirteen}`
      },
      icon: {
        primary: ebony.sixty
      },
      pagination: {
        bg: {
          primary: ebony.zero,
          secondary: ebony.twenty,
          tertiary: white.hundred,
          active: cornflower.fifty
        },
        text: {
          primary: ebony.seventy,
          secondary: white.hundred,
          active: white.hundred
        },
        border: {
          primary: 'none'
        }
      }
    }
  }),
  light: withCommons({
    text: {
      primary: ebony.zero,
      secondary: ebony.twenty,
      tertiary: cornflower.fifty,
      quaternary: ebony.seventy,
      disabled: ebony.sixty
    },
    bg: {
      primary: white.hundred,
      secondary: ebony.ninetySix,
      tertiary: ebony.ninety
    },
    icons: {
      primary: ebony.zero,
      secondary: cornflower.fifty
    },
    button: {
      bg: {
        primary: ebony.ninetySix,
        stroke: 'transparent',
        flat: 'transparent',
        rounded: white.hundred,
        disabled: ebony.ninetySix,
        group: 'transparent'
      },
      text: {
        primary: ebony.zero,
        stroke: ebony.sixty,
        flat: ebony.zero,
        group: ebony.eighty,
        rounded: cornflower.fifty,
        disabled: ebony.sixty
      },
      border: {
        primary: 'none',
        stroke: `1px solid ${ebony.sixty}`,
        flat: 'none',
        rounded: 'none',
        disabled: 'none',
        group: `1px solid ${ebony.eighty}`
      },
      hover: {
        primary: { bg: cornflower.fifty, text: white.hundred },
        stroke: {
          border: `1px solid ${cornflower.fifty}`,
          text: cornflower.fifty
        },
        flat: { bg: cornflower.fifty, text: white.hundred },
        group: { bg: ebony.sixty, text: white.hundred },
        rounded: { bg: white.hundred, text: cornflower.fifty }
      },
      active: {
        primary: cornflower.thirty,
        stroke: ebony.ninetySix,
        flat: cornflower.thirty,
        group: { bg: ebony.sixty, border: 'none' }
      }
    },
    slider: {
      bg: {
        on: cornflower.fifty,
        off: ebony.ninety,
        thumb: white.hundred
      },
      text: {
        primary: ebony.zero,
        secondary: cornflower.fifty
      },
      border: {
        thumb: `2px solid ${ebony.eighty}`
      }
    },
    drawer: {
      bg: {
        primary: white.hundred
      },
      text: {
        primary: ebony.zero,
        secondary: ebony.sixty
      },
      border: {
        primary: `1px solid ${ebony.ninety}`
      },
      icon: {
        primary: cornflower.fifty,
        secondary: ebony.zero
      },
      shadow: {
        primary: `0 4px 4px 0 ${shadow.default}`
      }
    },
    accordion: {
      bg: {
        primary: ebony.ninetySix
      },
      text: {
        primary: ebony.zero
      },
      icon: {
        primary: ebony.sixty,
        secondary: ebony.zero
      }
    },
    tabs: {
      text: {
        on: white.hundred,
        off: ebony.seventy
      },
      bg: {
        primary: 'transparent',
        secondary: cornflower.fifty
      }
    },
    selector: {
      text: {
        primary: ebony.zero,
        disabled: ebony.sixty
      },
      border: {
        on: `1px solid ${cornflower.fifty}`,
        off: `1px solid ${ebony.sixty}`,
        disabled: `1px solid ${ebony.eighty}`
      },
      bg: {
        on: cornflower.fifty,
        off: ebony.ninety
      },
      icon: {
        primary: white.hundred,
        disabled: ebony.eighty
      }
    },
    tooltip: {
      text: {
        primary: cornflower.fifty,
        secondary: white.hundred
      },
      bg: {
        primary: white.hundred,
        secondary: cornflower.fifty
      },
      header: {
        primary: ebony.ninetySix,
        secondary: ebony.twenty
      },
      shadow: {
        primary: `0 2px 4px 0 ${shadow.default}`
      }
    },
    snackbar: {
      bg: {
        primary: ebony.zero
      },
      text: {
        primary: white.hundred,
        secondary: ciano.fifty
      }
    },
    notification: {
      bg: {
        primary: ebony.thirteen
      },
      text: {
        primary: white.hundred
      },
      icon: {
        primary: white.hundred,
        secondary: ebony.zero
      },
      close: {
        primary: white.hundred
      },
      state: {
        primary: cornflower.fifty,
        danger: red.fifty,
        warning: yellow.fifty,
        success: green.fifty
      }
    },
    field: {
      bg: {
        primary: ebony.ninetySix,
        secondary: white.hundred,
        tertiary: ebony.eighty,
        active: cornflower.fifty,
        hover: cornflower.fifty
      },
      text: {
        primary: ebony.twenty,
        secondary: ebony.sixty,
        tertiary: ebony.twenty,
        quaternary: white.hundred,
        active: cornflower.fifty,
        selected: ebony.twenty,
        disabled: ebony.eighty,
        danger: red.fifty
      },
      border: {
        primary: `1px solid ${ebony.eighty}`,
        secondary: `1px solid ${ebony.sixty}`,
        tertiary: `1px solid ${ebony.ninetySix}`,
        danger: `1px solid ${red.fifty}`
      },
      shadow: {
        primary: `0 2px 4px ${shadow.default}`
      },
      arrow: {
        primary: ebony.sixty,
        selected: ebony.twenty
      },
      icon: {
        primary: ebony.sixty,
        secondary: white.hundred,
        tertiary: ebony.sixty,
        active: white.hundred,
        success: green.fifty
      }
    },
    datepicker: {
      text: {
        primary: ebony.zero,
        secondary: white.hundred,
        tertiary: ebony.sixty,
        off: ebony.seventy
      },
      bg: {
        primary: white.hundred,
        secondary: cornflower.fifty,
        tertiary: ebony.ninetySix
      },
      icon: {
        primary: cornflower.fifty,
        secondary: white.hundred,
        tertiary: ebony.sixty
      },
      shadow: {
        primary: `0 2px 4px ${shadow.default}`
      },
      border: {
        primary: `1px solid ${ebony.ninety}`
      }
    },
    navigation: {
      bg: {
        primary: ebony.ninetySix,
        secondary: ebony.ninety,
        tertiary: ebony.eighty,
        hover: white.hundred
      },
      separator: {
        primary: ebony.ninety
      },
      text: {
        primary: ebony.twenty,
        hover: cornflower.fifty
      },
      shadow: {
        primary: `0 0 16px ${shadow.darker}`
      }
    },
    menu: {
      bg: {
        primary: ebony.ninetySix
      },
      toggle: {
        primary: white.hundred,
        secondary: ebony.ninety,
        tertiary: cornflower.fifty
      }
    },
    player: {
      text: {
        primary: white.hundred,
        secondary: ebony.zero
      },
      bg: {
        primary: ebony.zero,
        secondary: white.hundred,
        opacity: '0.8'
      },
      icon: {
        primary: white.hundred,
        secondary: ebony.zero
      },
      progress: {
        on: white.hundred,
        off: gray.fifty
      },
      shadow: {
        primary: `0px 0px 4px 0px ${shadow.default}`
      }
    },
    datatable: {
      text: {
        primary: ebony.zero,
        hover: ebony.zero
      },
      bg: {
        primary: white.hundred,
        rowHover: ebony.ninetySix,
        columnHover: ebony.ninetySix
      },
      border: {
        primary: `1px solid ${ebony.ninety}`
      },
      icon: {
        primary: ebony.sixty
      },
      pagination: {
        bg: {
          primary: white.hundred,
          secondary: ebony.ninetySix,
          tertiary: white.hundred,
          active: cornflower.fifty
        },
        text: {
          primary: ebony.seventy,
          secondary: cornflower.fifty,
          active: white.hundred
        },
        border: {
          primary: `1px solid ${ebony.ninety}`
        }
      }
    }
  })
}
