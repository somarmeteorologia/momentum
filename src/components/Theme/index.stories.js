// import React from 'react'
// import { storiesOf } from '@storybook/react'

// import { GROUPS } from '@helpers/hierarchySeparators'

// import { Reset } from '@components/Reset'

// // storiesOf(`${GROUPS.TOKENS}|Color`, module).addDecorator(story => (
// //   <>
// //     <Reset />
// //     {story()}
// //   </>
// // ))
// // .add('Default', () => <Colors colors={colors} />)
