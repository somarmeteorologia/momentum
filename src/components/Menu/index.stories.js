import React, { useState } from 'react'
import { storiesOf } from '@storybook/react'
import styled from 'styled-components'

import { GROUPS } from '@helpers/hierarchySeparators'

import { Reset } from '@components/Reset'
import { Menu as Menuable } from '@components/Menu'

const Content = styled.div`
  width: 100%;
`

const Container = styled.div`
  display: flex;
`

const Menu = ({ draggable, children, width }) => {
  const [isOpen, setOpen] = useState(true)

  return (
    <Menuable
      draggable={draggable}
      isOpen={isOpen}
      setOpen={setOpen}
      width={width}
    >
      {children}
    </Menuable>
  )
}

storiesOf(`${GROUPS.COMPONENTS}|Menu`, module)
  .addDecorator(story => (
    <>
      <Reset />
      {story()}
    </>
  ))
  .add('Default', () => (
    <Container>
      <Menu>
        <div>content</div>
      </Menu>

      <Content />
    </Container>
  ))
  .add('Dragging', () => (
    <>
      <Menu draggable={true}>
        <div>content</div>
      </Menu>

      <Content />
    </>
  ))
  .add('Custom width', () => (
    <>
      <Menu width={350}>
        <div>content</div>
      </Menu>

      <Content />
    </>
  ))
