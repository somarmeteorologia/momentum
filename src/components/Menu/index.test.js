import React, { useState } from 'react'
import { render, fireEvent, wait } from '@testing-library/react'
import { ThemeProvider } from 'styled-components'

import { Theme } from '@components/Theme'
import { Menu as Menuable } from '@components/Menu'

const Menu = ({ draggable }) => {
  const [isOpen, setOpen] = useState(true)

  return <Menuable draggable={draggable} isOpen={isOpen} setOpen={setOpen} />
}

const withProvider = ({ children, ...props }) => (
  <ThemeProvider theme={Theme.light}>
    <Menu {...props}>{children}</Menu>
  </ThemeProvider>
)

const INTERVAL = { interval: 300 }

test('Not expect visual regression', () => {
  const { container } = render(withProvider({}))

  expect(container.firstChild).toMatchSnapshot()
})

test('Expect show and hide menu button must be working', async () => {
  const { getByTestId } = render(
    withProvider({
      children: <div data-testid="content">content</div>
    })
  )

  expect(getByTestId('menu')).toHaveStyle('width: 300px')

  fireEvent.mouseEnter(getByTestId('menu'))
  fireEvent.click(getByTestId('toggleable'))

  await wait(() => {
    expect(getByTestId('menu')).toHaveStyle('width: 20px')
  }, INTERVAL)

  fireEvent.mouseEnter(getByTestId('menu'))
  fireEvent.click(getByTestId('toggleable'))

  await wait(() => {
    expect(getByTestId('menu')).toHaveStyle('width: 300px')
  }, INTERVAL)
})
