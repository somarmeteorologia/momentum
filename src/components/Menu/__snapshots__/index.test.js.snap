// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Not expect visual regression 1`] = `
.c0 {
  height: 100vh;
  background-color: #f1f3fA;
  position: relative;
  display: block;
  box-shadow: 0 0 16px rgba(50,71,151,0.25);
}

<div
  class="c0"
  data-testid="menu"
  style="width: 300px;"
>
  <div
    class=""
    style="display: block; opacity: 1; transform: none;"
  />
</div>
`;
